### ISSUE TITLE
[parallel-solving] segmentation fault

### ISSUE THREAD

**Issue by lucasccordeiro** [esbmc/esbmc#1028]
ESBMC produces a segmentation fault if we run the following command:

 esbmc [main.c](https://github.com/esbmc/esbmc/blob/master/regression/esbmc-unix/05_pfscan-1.0_01/main.c) --incremental-bmc --32 --color --parallel-solving

````
Slicing for Claim dereference failure: Incorrect alignment when accessing data object (0.000s)
Slicing time: 0.007s (removed 444 assignments)
No solver specified; defaulting to Boolector
Slicing time: 0.008s (removed 444 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Segmentation fault
````

**Comment by ChenfengWei0** [comment]
Hmm... I had the #1011 problem when running this test, with or without `--parallel-solving`

```sh
esbmc main.c --incremental-bmc --32

...
ERROR: Looking up index of nonexistant member "__lock" in struct/union "pthread_mutex_t"
[1]    6092 IOT instruction  ../../../../release/bin/esbmc main.c --incremental-bmc --32
```

I think we should also test the other two options (`--multi-property` and `--multi-fail-fast`) , to see if they work properly

**Comment by fbrausse** [comment]
There is a race condition when hashing expr2t's for the `smt_cache` in smt_conv. Most likely the reason is that all parallel jobs refer to the same expr2tc objects as members of `symex_target_equationt::SSA_stept`.

*Edit:* for debugging this particular example for now you can use #823. The race condition however exists independently of whether the program uses pthreads.

**Comment by ChenfengWei0** [comment]
> There is a race condition when hashing expr2t's for the `smt_cache` in smt_conv. Most likely the reason is that all parallel jobs refer to the same expr2tc objects as members of `symex_target_equationt::SSA_stept`.
> 
> _Edit:_ for debugging this particular example for now you can use #823. The race condition however exists independently of whether the program uses pthreads.

Thanks for your help!

**Comment by ChenfengWei0** [comment]
Based on the #832 
- `--multi-property` (the non-parallel-solving) works fine
- `--parallel-sovling` raises assertion failure

```sh
esbmc: /root/workspace/ESBMC_Commit/esbmc/src/util/string_container.h:68: const std::string &string_containert::get_string(size_t) const: Assertion `no < string_vector.size()' failed.
esbmc: /root/workspace/ESBMC_Commit/esbmc/src/util/string_container.h:68: const std::string &string_containert::get_string(size_t) const: Assertion `no < string_vector.size()' failed.
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
esbmc: /root/workspace/ESBMC_Commit/esbmc/src/util/string_container.h:68: const std::string &string_containert::get_string(size_t) const: Assertion `no < string_vector.size()' failed.
esbmc: /root/workspace/ESBMC_Commit/esbmc/src/util/string_container.h:68: const std::string &string_containert::get_string(size_t) const: Assertion `no < string_vector.size()' failed.
esbmc: /root/workspace/ESBMC_Commit/esbmc/src/util/string_container.h:68: const std::string &string_containert::get_string(size_t) const: Assertion `no < string_vector.size()' failed.
```

**Comment by fbrausse** [comment]
> * `--parallel-sovling` raises assertion failure

Yeah, I guess these are not the only ones. All global caches are potentially subject to these kinds of races. One way out might be locking modified shared state (such as the `expr2t::crc_val`, the `string_containert`, etc.) with mutexes or as atomics, though I suspect that for the main use case of caching this has too much overhead. Another way might be to make those things thread-local.

**Comment by fbrausse** [comment]
There clearly are a lot of caches (both global and local) which lead to race conditions with the threaded parallel-solving. Additionally the parallel-solving code in the multi_property_check() is pretty hard to read.

@rafaelsamenezes @ChenfengWei0 Do you know whether we have any users of the parallel-solving? Would you consider removing the multi-threading to clean up this code?

**Comment by rafaelsamenezes** [comment]
> @rafaelsamenezes @ChenfengWei0 Do you know whether we have any users of the parallel-solving? Would you consider removing the multi-threading to clean up this code?

Only as an experimental feature. I am fine with removing it for a cleaning.

**Comment by lucasccordeiro** [comment]
This issue is not valid anymore since we removed the support for parallel solving.
